tools/qfvis/source/flow.c

```diff
@@ -300,6 +300,234 @@ free_winding_memory (threaddata_t *thread, size_t winding_mark)
 	Hunk_RawFreeToLowMark (thread->hunk, winding_mark);
 }
 
+/*
+	TargetChecks
+
+	Filter mightsee by clipping against all portals
+*/
+static unsigned
+TargetChecks(threaddata_t *thread,
+			 pstack_t *prevstack,
+			 set_t **mainportalbits,
+			 set_t **tempportalbits)
+{
+	unsigned    i, initial_portalcheck;
+	portal_t   *target_portal;
+	vec4f_t     backplane;
+	vec4f_t     source_plane, pass_plane;
+	const winding_t *pass_winding;
+	winding_t  *source_winding, *target_winding;
+
+	if (prevstack->pass_winding == NULL)
+		return 0;
+	
+	thread->stats.chains++;
+
+	// initialize counter
+	initial_portalcheck = thread->stats.portalcheck;
+
+	// reuse separators
+	sep_t *separators[2] = {NULL, NULL};
+	if (prevstack->next) {
+		separators[0] = prevstack->next->separators[0];
+		separators[1] = prevstack->next->separators[1];
+	}
+
+	// empty cluster set
+	set_t *mightsee = thread->scratchbits;
+	set_empty(mightsee);
+
+	// empty portal set
+	set_t *prevportalbits = *mainportalbits;
+	set_t *portalbits = *tempportalbits;
+	set_empty(portalbits);
+	*mainportalbits = portalbits;
+	*tempportalbits = prevportalbits;
+
+	source_plane = thread->pstack_head.pass_plane;
+	pass_winding = prevstack->pass_winding;
+	pass_plane = prevstack->pass_plane;
+
+	// lower than 2 won't have a benefit. higher than 2 refines the source
+	// portal even though it isn't propagated forward right now.
+	int level = 2; // = options.level
+
+	size_t      winding_mark = Hunk_LowMark (thread->hunk);
+
+	// check all portals for flowing into other leafs
+	for (i = 0, target_portal = portals; i < numportals * 2; i++, target_portal++) {
+
+		if (!set_is_member(prevstack->mightsee, target_portal->cluster))
+            continue;           // can't possibly see it
+		if (!set_is_member(prevportalbits, i))
+            continue;           // can't possibly see it
+		if (set_is_member(mightsee, target_portal->cluster)) {
+			// target check already done and passed
+			set_add(portalbits, i); // let the portal survive to the next level
+            continue;           
+		}
+
+		// get plane of target_portal, point normal into the neighbor cluster
+		backplane = -target_portal->plane;
+
+		vec4f_t     diff = vabs4f (pass_plane - backplane);
```

Comment:
Is this block correct? It's checking only that the normals are close to (anti-)parallel and ignoring the plane distances thus there's no guarantee the planes are actually co-planar.

Reply:
I double checked, and a proper co-planar check looks like:
```c
vec4f_t dir = crossf (pass_plane, backplane);
vec4f_t pos = pass_plane[3] * backplane - backplane[3] * pass_plane;
// co-planar if both dir and pos (components 0..2, ignore 3) are (near) zero
```

Though I suspect the `abs` check you have is good enough for the `dir` vector.

For reference, a plane's vec4 is `{ n_x, n_y, n_z, -n•p }` (which is why dotting a plane's vec4 with a point's (homogeneous) vec4 works to give point-plane distance). Also, `dir` and `pos` together form the 6d Plucker coordinates of the line of intersection of the two planes, even if they are parallel.

However, I am unsure what your intention is here; a more detailed comment would do wonders.